### PR TITLE
Improves checkpointerWriter memory usage

### PR DESCRIPTION
**Pull request by cyriltovena**
I've improve marshalling and pooling to reduce memory usage. I've also fixes a small mistake (`1>>20 =0`).

I've also added some more debug logs, I want to see how many each series weight and at what size do we actually flush them.

benchmp:

```
> benchcmp  before.txt after.txt4
benchmark                        old ns/op     new ns/op     delta
Benchmark_CheckpointWrite-16     5218516       1046717       -79.94%

benchmark                        old allocs     new allocs     delta
Benchmark_CheckpointWrite-16     165            220            +33.33%

benchmark                        old bytes     new bytes     delta
Benchmark_CheckpointWrite-16     31645483      23673         -99.93%
```

### CODE REVIEW

File: pkg/ingester/checkpoint.go [hunk 1]
```diff
@@ -260,30 +273,40 @@ func (w *WALCheckpointWriter) Advance() (bool, error) {
...
 		return err
 	}
 
 	w.recs = append(w.recs, b)
 	w.bufSize += len(b)
+	level.Debug(util.Logger).Log("msg", "writing series", "size", humanize.Bytes(uint64(len(b))))
 
 	// 1MB
-	if w.bufSize > 1>>20 {
+	if w.bufSize > 1<<20 {
```
Comment (cyriltovena) [inline pkg/ingester/checkpoint.go:293]: @owen-d :)

**Review by sandeepsukhani — APPROVED** [review]
LGTM

File: pkg/ingester/checkpoint.go (hunk 1 above)
Comment (owen-d) [reply pkg/ingester/checkpoint.go:293]: ughh

File: pkg/ingester/checkpoint.go [hunk 2]
```diff
@@ -96,16 +98,17 @@ func decodeCheckpointRecord(rec []byte, s *Series) error {
...
 }
 
-func encodeWithTypeHeader(m proto.Message, typ RecordType) ([]byte, error) {
-	buf, err := proto.Marshal(m)
+func encodeWithTypeHeader(m *Series, typ RecordType, buf []byte) ([]byte, error) {
+	size := m.Size()
+	if cap(buf) < size+1 {
+		buf = make([]byte, size+1)
+	}
+	_, err := m.MarshalTo(buf[1 : size+1])
```
Comment (owen-d) [inline pkg/ingester/checkpoint.go:106]: <3 @cyriltovena

**Review by owen-d — APPROVED** [review]

**Pull request merged**